### PR TITLE
Add A/B test to allow us to turn off background compilation/parsing

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
I'm pretty skeptical that this system serves any purpose.  We already getting features calling us on extremely fast cadences, so my expectation is that we're already pulling things like the trees/compilations forward in VS.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski ptal.

**Review by jasonmalinowski — APPROVED** [review]
Seems reasonable.

I suspect we're not going to see a huge change here. We definitely have plenty of features pulling on things to make progress, so I'm not worried about that never happening. If there's a benefit, I imagine it means that a feature that triggers on a tiny delay happens just a bit sooner. So if tags start computing 50ms after a keystroke, before the tree/compilation likely would be there, but now we'll be including that time in the operation. We might see an uptick in things like LSP response times a result. Now is that _bad_? I don't know, but I'd say communicate when you turn this on just so if we do see a change in data we know this _might_ be a cause.

File: src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspace.cs [hunk 1]
```diff
@@ -32,44 +34,57 @@ static VisualStudioWorkspace()
...
 
         protected internal override bool PartialSemanticsEnabled
         {
-            get { return _backgroundCompiler != null; }
+            get
+            {
+                // Still allow for partial semantics in the VS workspace, even with background compilation disabled. The
+                // presumption here is that there's *always* something implicitly pulling the compilations forward in VS
+                // (e.g. tagging, completion, etc.) at a regular cadence.
+                if (_backgroundCompilationDisabled)
```
Comment (genlu) [inline src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspace.cs:76]: Would requesting a frozen partial solution also pulling the complilations forward?

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspace.cs:76]: no.  frozen partial is specifically around the idea of saying "i want a snapshot of whatever you have now, i don't want to wait for all information to be up to date.  but make sure you at least have the contents of this documetn correct as i'm going to ask about the tree/semantics for *it*".

Comment (genlu) [reply src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspace.cs:76]: ok, then we need to remove "completion" from the features that pull us forward in the comment :)

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspace.cs:76]: fair.  will update :)

File: src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspace.cs [hunk 2]
```diff
@@ -32,44 +34,57 @@ static VisualStudioWorkspace()
...
         }
 
         protected internal override bool PartialSemanticsEnabled
         {
-            get { return _backgroundCompiler != null; }
+            get
+            {
+                // Still allow for partial semantics in the VS workspace, even with background compilation disabled. The
+                // presumption here is that there's *always* something implicitly pulling the compilations forward in VS
+                // (e.g. tagging, completion, etc.) at a regular cadence.
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspace.cs:75]:
```suggestion
                // (e.g. tagging, lightbulb, etc.) at a regular cadence.
```

File: src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspace.cs [hunk 3]
```diff
@@ -32,44 +34,57 @@ static VisualStudioWorkspace()
         internal VisualStudioWorkspace(HostServices hostServices)
             : base(hostServices, WorkspaceKind.Host)
         {
-            _backgroundCompiler = new BackgroundCompiler(this);
-
-            var cacheService = Services.GetService<IWorkspaceCacheService>();
-            if (cacheService != null)
+            _backgroundCompilationDisabled = this.Services.GetRequiredService<IWorkspaceConfigurationService>().Options.DisableBackgroundCompilation;
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspace.cs:37]: ❔ Are feature flags updated by this point?

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspace.cs:37]: It's all done async by the platform. So we will read the value as is at startup. Then, when it changes we will get the fresh value on next session.

**Pull request merged**